File: examples/js/character_data.js
#!/usr/bin/gjs

const GXml = imports.gi.GXml;

print ("CharacterData isn't directly instantiable, but its subclasses Text and Comment are!\n");

// Setup
var doc = GXml.GomDocument.from_path ("bookshelf_node.xml");
var books_list = doc.document_element.get_elements_by_tag_name ("Book");
var book = books_list.get_element (0);

// Create a Text node for CharacterData
var text = doc.create_text_node ("Some infinities are bigger than other infinities");
var comment = doc.create_comment ("As he read, I fell in love the way you fall asleep:");

/* Data representation */

print ("Stringifying Text:\n" +
       text.data + "\n\n" +
       "Text's CharacterData's data:\n" +
       text.data + "\n\n" +
       "Text's CharacterData's length:\n" +
       text.get_length () + "\n\n" +
       "Stringifying Comment:\n" +
       comment.data + "\n\n" +
       "Comment's CharacterData's data:\n" +
       comment.data + "\n\n" +
       "Comment's CharacterData's length:\n" +
       comment.get_length () + "\n\n");

/* CharacterData operations */

book.append_child (text);
print ("Book with our Text as its child (its content):\n" + book.write_string ());

book.replace_child (comment, text);
print ("\nBook with our Comment as its child (its content):\n" + book.write_string ());

comment.append_data (" slowly, and then all at once");
print ("\nBook after appending more data to Comment:\n" + book.write_string ());

comment.replace_data (3, 2, "Gus");
print ("\nBook after replacing 'he' with 'Gus':\n" + book.write_string ());

comment.delete_data (20, 8);
print ("\nBook after deleting 'in love ':\n" + book.write_string ());

comment.insert_data (20, "in love ");
print ("\nBook after inserting 'in love ':\n" + book.write_string ());

/* You can see some more examples by looking at Comment and Text's examples */

File: examples/c/character_data.c
#include <gxml/gxml.h>

int main () {
  /* Setup */
  printf ("CharacterData isn't directly instantiable, but its subclasses Text and Comment are!  Go see them for more additional examples.\n");

  GXmlDocument *doc;
  GXmlNodeList *books_list;
  GXmlNode *book;

  doc = gxml_document_new_from_path ("bookshelf_node.xml");
  books_list = gxml_element_get_elements_by_tag_name (gxml_document_get_document_element (doc), "Book");
  book = gxml_node_list_nth (books_list, 0);

  GXmlCharacterData *text;
  GXmlCharacterData *comment;
  text = GXML_CHARACTER_DATA (gxml_document_create_text_node (doc, "Some infinities are bigger than other infinities"));
  comment = GXML_CHARACTER_DATA (gxml_document_create_comment (doc, "As he read, I fell in love the way you fall asleep:"));

  gchar *str;
  gchar *str2;
  
  /* Data representation */

  printf ("Stringifying Text:\n%s\n\n"
	  "Text's CharacterData's data:\n%s\n\n"
	  "Text's CharacterData's length:\n%lu\n\n"
	  "Stringifying Comment:\n%s\n\n"
	  "Comment's CharacterData's data:\n%s\n\n"
	  "Comment's CharacterData's length:\n%lu\n\n",
	  str = gxml_node_to_string (GXML_NODE (text), TRUE, 0),
	  gxml_character_data_get_data (text),
	  gxml_character_data_get_length (text),
	  str2 = gxml_node_to_string (GXML_NODE (comment), TRUE, 0),
	  gxml_character_data_get_data (comment),
	  gxml_character_data_get_length (comment));
  g_free (str);
  g_free (str2);
	  
  /* CharacterData operations */

  gxml_node_append_child (book, GXML_NODE (text));
  printf ("Book with our Text as its child (its content):\n%s\n\n",
	  str = gxml_node_to_string (book, TRUE, 0));
  g_free (str);

  gxml_node_replace_child (book, GXML_NODE (comment), GXML_NODE (text));
  printf ("Book with our Comment as its child (its content):\n%s\n\n",
	  str = gxml_node_to_string (book, TRUE, 0));
  g_free (str);
  
  gxml_character_data_append_data (comment, " slowly, and then all at once");
  printf ("Book after appending more data to Comment:\n%s\n\n",
	  str = gxml_node_to_string (book, TRUE, 0));
  g_free (str);
  
  gxml_character_data_replace_data (comment, 3, 2, "Gus");
  printf ("Book after replacing 'he' with 'Gus':\n%s\n\n",
	  str = gxml_node_to_string (book, TRUE, 0));
  g_free (str);
  
  gxml_character_data_delete_data (comment, 20, 8);
  printf ("Book after deleting 'in love ':\n%s\n\n",
	  str = gxml_node_to_string (book, TRUE, 0));
  g_free (str);

  gxml_character_data_insert_data (comment, 20, "in love ");
  printf ("Book after inserting 'in love ':\n%s\n",
	  str = gxml_node_to_string (book, TRUE, 0));
  g_free (str);  
 
  g_object_unref (doc);
  return 0;
}
